wgengine/netstack/netstack.go

```diff
@@ -302,6 +352,74 @@ func (ns *Impl) wrapProtoHandler(h func(stack.TransportEndpointID, stack.PacketB
 	}
 }
 
+// wrapTCPProtocolHandler wraps the protocol handler we pass to netstack for TCP.
+func (ns *Impl) wrapTCPProtocolHandler(h protocolHandlerFunc) protocolHandlerFunc {
+	return func(tei stack.TransportEndpointID, pb stack.PacketBufferPtr) (handled bool) {
+		localIP, ok := netip.AddrFromSlice(tei.LocalAddress.AsSlice())
+		if !ok {
+			ns.logf("netstack: could not parse local address for incoming connection")
+			return false
+		}
+		localIP = localIP.Unmap()
+
+		remoteIP, ok := netip.AddrFromSlice(tei.RemoteAddress.AsSlice())
+		if !ok {
+			ns.logf("netstack: could not parse remote address for incoming connection")
+			return false
+		}
+
+		// If we have too many pending connections for this client, abort
+		// early and don't open a new one.
+		//
+		// NOTE: the counter is decremented in
+		// decrementPendingTCPForward, called from the acceptTCP
+		// function, below.
+		ns.mu.Lock()
+		pending := ns.connsPendingByClient[remoteIP]
+		tooManyPending := pending > maxInFlightConnectionAttemptsPerClient()
+		if !tooManyPending {
+			ns.connsPendingByClient[remoteIP]++
+		}
+		ns.mu.Unlock()
+		if debugNetstack() {
+			ns.logf("[v2] netstack: pending connections for client %v: %d", remoteIP, pending)
+		}
+		if tooManyPending {
+			return false // unhandled
+		}
+
+		// On return, if this packet isn't handled, we need to
+		// decrement the per-client pending count. This can happen if
```

Comment:
Incrementing `connsPendingByClient` is conditional on the same `tooManyPending` bool we use to return `false`; do we really need to decrement here?

Reply:
Yes, though it's a bit nuanced; if the "inner" handler returns `false` to indicate that the packet is not handled, our `acceptTCP` function would never run and thus never decrement the per-client count; let me clarify the comment.